src/lib/support/BufBound.h

```diff
@@ -0,0 +1,137 @@
+/*
+ *
+ *    Copyright (c) 2020 Project CHIP Authors
+ *
+ *    Licensed under the Apache License, Version 2.0 (the "License");
+ *    you may not use this file except in compliance with the License.
+ *    You may obtain a copy of the License at
+ *
+ *        http://www.apache.org/licenses/LICENSE-2.0
+ *
+ *    Unless required by applicable law or agreed to in writing, software
+ *    distributed under the License is distributed on an "AS IS" BASIS,
+ *    WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ *    See the License for the specific language governing permissions and
+ *    limitations under the License.
+ */
+
+/**
+ *  @file
+ *    basic double-ended queue, intended to be embedded as a member
+ *     of an object
+ */
+
+#ifndef CHIP_BUFBOUND_H
+#define CHIP_BUFBOUND_H
+
+#include <stdint.h>
+#include <string.h>
+
+namespace chip {
+
+/**
+ *  @class BufBound
+ *
+ *  simple buffer writer
+ */
+class BufBound
+{
+private:
+    uint8_t * mBuf;
+    size_t mLen;
+    size_t mWritten;
+
+public:
+    BufBound(uint8_t * buf, size_t len) : mBuf(buf), mLen(len), mWritten(0) {}
+
+    /*
+     * @brief append a single byte
+     */
+    size_t Put(uint8_t c)
```

Comment:
Why not implement this based on `Put(const uint8_t * buf, size_t len)` with `Put(&c, sizeof (c))`?

Reply:
less code to do it inline, I think

Reply:
Can we quantify that?

Reply:
If that is a gate for this PR, sure.  Is it?

Reply:
thought it might be interesting to explore, so:
**BufBound.h**
```
    /*
     * @brief append a single byte
     */
    size_t Put(uint8_t c)
    {
#ifdef USE_MEMMOVE
        return Put(&c, sizeof(c));
#else
        if (mWritten < mLen)
        {
            mBuf[mWritten] = c;
        }
        return ++mWritten;
#endif
    }
```
**BufBoundPut.cpp**
```
#include "BufBound.h"

void Put()
{
    chip::BufBound bb(nullptr, 0);
    uint8_t c = 1;
    bb.Put(c);
}
```
size with USE_MEMMOVE:
```
-*- mode: compilation; default-directory: "~/dev/proto/C/" -*-
Compilation started at Mon Aug  3 10:08:13

~/.espressif/tools/xtensa-esp32-elf/esp-2019r2-8.2.0/xtensa-esp32-elf/bin/xtensa-esp32-elf-g++ -DUSE_MEMMOVE -c BufBoundPut.cpp && ~/.espressif/tools/esp32ulp-elf/2.28.51.20170517/esp32ulp-elf-binutils/bin/esp32ulp-elf-size BufBoundPut.o
   text	   data	    bss	    dec	    hex	filename
    316	      0	      0	    316	    13c	BufBoundPut.o
```
size without:
```
-*- mode: compilation; default-directory: "~/dev/proto/C/" -*-
Compilation started at Mon Aug  3 10:08:43

~/.espressif/tools/xtensa-esp32-elf/esp-2019r2-8.2.0/xtensa-esp32-elf/bin/xtensa-esp32-elf-g++ -c BufBoundPut.cpp && ~/.espressif/tools/esp32ulp-elf/2.28.51.20170517/esp32ulp-elf-binutils/bin/esp32ulp-elf-size BufBoundPut.o
   text	   data	    bss	    dec	    hex	filename
    126	      0	      0	    126	     7e	BufBoundPut.o

Compilation finished at Mon Aug  3 10:08:43
```

Reply:
Thanks for poking at that. Compiled with -Os? If so, I'm surprised that the compiler can't optimize better than that, particularly with a header implementation (speaking of which, since this isn't a template, any reason not to put the implementation in a source file?

Reply:
yeah, things change when optimization is introduced.

... more hacking ...

Conclusions:

With -Os and -O2, thumb-code text is **_5 bytes_** whether or not the inline version is included (i.e. the inline form doesn't cost anything).

With optimizations on, runtime performance of the inline form is between 4X and 6X faster than calling memmove() for a single byte.